pkg/receive/handler.go

```diff
@@ -358,82 +366,116 @@ func (h *Handler) parallelizeRequests(ctx context.Context, tenant string, replic
 		}
+
 		// Make a request to the specified endpoint.
 		go func(endpoint string) {
-			var err error
+			defer wg.Done()
 
-			// Increment the counters as necessary now that
-			// the requests will go out.
+			var (
+				err error
+				cl  storepb.WriteableStoreClient
+			)
 			defer func() {
+				// This is an actual remote forward request so report metric here.
 				if err != nil {
 					h.forwardRequestsTotal.WithLabelValues("error").Inc()
 					return
 				}
 				h.forwardRequestsTotal.WithLabelValues("success").Inc()
 			}()
 
-			cl, err := h.peers.get(ctx, endpoint)
+			cl, err = h.peers.get(ctx, endpoint)
 			if err != nil {
-				level.Error(h.logger).Log("msg", "failed to get peer connection to forward request", "err", err, "endpoint", endpoint)
-				ec <- err
+				ec <- errors.Wrapf(err, "get peer connection for endpoint %v", endpoint)
 				return
 			}
 			// Create a span to track the request made to another receive node.
 			tracing.DoInSpan(ctx, "receive_forward", func(ctx context.Context) {
-				// Actually make the request against the endpoint
-				// we determined should handle these time series.
+				// Actually make the request against the endpoint we determined should handle these time series.
 				_, err = cl.RemoteWrite(ctx, &storepb.WriteRequest{
 					Timeseries: wreqs[endpoint].Timeseries,
 					Tenant:     tenant,
-					Replica:    int64(replicas[endpoint].n + 1), // increment replica since on-the-wire format is 1-indexed and 0 indicates unreplicated.
+					// Increment replica since on-the-wire format is 1-indexed and 0 indicates un-replicated.
+					Replica: int64(replicas[endpoint].n + 1),
 				})
-				if err != nil {
-					level.Error(h.logger).Log("msg", "forwarding request", "err", err, "endpoint", endpoint)
-					ec <- err
-					return
-				}
-				ec <- nil
 			})
+			if err != nil {
+				ec <- errors.Wrapf(err, "forwarding request to endpoint %v", endpoint)
+				return
+			}
+			ec <- nil
 		}(endpoint)
 	}
 
-	// Collect any errors from forwarding the time series.
-	// Rather than doing a wg.Wait here, we decrement a counter
-	// for every error received on the chan. This simplifies
-	// error collection and avoids data races with a separate
-	// error collection goroutine.
-	var errs terrors.MultiError
-	for ; n > 0; n-- {
-		if err := <-ec; err != nil {
-			errs.Add(err)
+	go func() {
+		wg.Wait()
+		close(ec)
+	}()
+
+	// At the end, make sure to exhaust the channel, letting remaining unnecessary requests finish asnychronously.
+	// This is needed if context is cancelled or if we reached success of fail quorum faster.
+	defer func() {
+		go func() {
+			for err := range ec {
+				if err != nil {
+					level.Debug(h.logger).Log("msg", "request failed, but not needed to achieve quorum", "err", err)
+				}
+			}
+		}()
+	}()
+
+	var (
+		success int
+		errs    terrors.MultiError
+	)
+	for {
+		select {
+		case <-ctx.Done():
+			return ctx.Err()
+		case err, more := <-ec:
+			if !more {
+				return errs
```

Comment:
I was thinking about this case, but I though that this will never happen (: We either have success or errors quorum kind of in my previous version (:

Reply:
The attempt here is to return the best possible error, for example in a 3x replication, and 2x return tsdb-not-ready/unavailable, and 1 conflict, would end up with a generalized error when in reality a retry is likely to resolve the error.